docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,408 @@
+consumes their catalog entry.
+
+## Prior Art
+
+### Helm
+
+[Helm](https://github.com/kubernetes/helm) uses “charts” (e.g. resource
+templates) that can instantiated as “releases”. Charts are stored in a HTTP
+repository as compressed tarballs.  The tarball contains, among other files:
+
+- `chart.yaml` - metadata about the chart
+- `values.yaml` -  default variable values to use when processing the resource
+   template files
+- `templates/` - Kubernetes resource templates processed at `helm install` time
+
+Workflow:
+
+- `helm init` -  starts Tiller in the cluster and initializes CLI environment.
+   Tiller runs in the cluster and manages the running Releases.
+- `helm install <chartname>` - downloads the chart, processes the resource
+   templates, and instantiates the resources in a release
+
+Helm addresses use case 2.1: it provides a way to both create templates and
+search for/consume them.  Charts can also incorporate secrets, however, secrets
+created as part of a Helm release are not updated when the secret in the chart
+is modified.
+
+It is also possible to address use case 4.2; a user could take advantage of
+Helm’s support for pre hooks to e.g. run a job that requests that a new
+username/password be provisioned. The job, however, would need to have
+sufficient privileges to create a username/password, which means the
+administrative credentials necessary to provision the new account would need to
+be included in the chart itself, which is a security risk and presumably not a
+viable solution. Alternatively, the job would need to make an unprivileged
+request to some other service that does have sufficient privileges.
+
+TODO
+Also: 7.1, use pre-hooks to provision resources
+
+### Docker Compose
+
+Docker Compose is most like a pod spec file.  It can contain information about
+container(s) to be deployed together (similar to a pod) and the volumes,
+environment variables, and ports that the combined composition will use.
+
+Compose is similar to Helm in that it can bring up and take down a templated set
+of resources as a unit.  Obviously, Compose does not create Kubernetes resources
+as Helm does, including Services which make the service offered by the
+composition discoverable.
+
+### OpenShift Templates
+
+Openshift templates offer a feature set similar to Helm.  Templates are
+resources created as yaml files and imported as a resource of type "Template"
+into Openshift.  Templates are processed with the "oc process" command, filling
+in fields from parameter list and can even dynamically generate parameters like
+passwords, for example. Once processed, the template becomes a single yaml file
+output that can be fed to `oc create -f` to create the resources in Openshift.
+
+The use case application is the same as that of Helm.  Once resources created by
+the template, they are not tracked any further as a member of the template.  It
+is up to the users to apply labels such that all resources created by a template
```

Comment:
It would be useful to be able to apply labels more generically. See:
https://github.com/kubernetes/kubernetes/blob/master/docs/design/metadata-policy.md
#17097
#1698

Note that adding labels isn't sufficient to avoid selector overlap. We'd also need selector generation #12298 and/or controllerRef #24946.